File: cplusplus/44WildcardMatching.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        vector<vector<bool>> matchingTable(s.size() + 1, vector<bool>(p.size() + 1, false));
        matchingTable[0][0] = true;
        
        for (int j = 0; j < p.size(); ++j) {
            if (p.at(j) == '*') {
                matchingTable[0][j + 1] = matchingTable[0][j];
            }
        }
        
        for (int i = 0; i < s.size(); ++i) {
            for (int j = 0; j < p.size(); ++j) {
                if (p.at(j) != '*') {
                    if (p.at(j) == '?' || s.at(i) == p.at(j))
                        matchingTable[i+1][j+1] = matchingTable[i][j];
                    else matchingTable[i+1][j+1] = false;
                } else {
                    if (matchingTable[i+1][j] || matchingTable[i][j+1])
                        matchingTable[i+1][j+1] = true;
                    else matchingTable[i+1][j+1] = false;
                }
            }
        }
        
        return matchingTable[s.size()][p.size()];
    }
};

File: java/44WildcardMatching.java
class Solution {
    public boolean isMatch(String s, String p) {        
        boolean[][] res = new boolean[s.length() + 1][p.length() + 1];
        
        for (int i = s.length(); i >= 0; --i) {
            for (int j = p.length(); j >= 0; --j) {
                if (i == s.length() && j == p.length()) {
                    res[i][j] = true;
                } else if (j == p.length()) {
                    res[i][j] = false;
                } else if (i == s.length()) {
                    char c = p.charAt(j);
                    if (c != '*')
                        res[i][j] = false;
                    else res[i][j] = res[i][j + 1];
                } else {
                    char c1 = s.charAt(i);
                    char c2 = p.charAt(j);
                
                    if (c1 == c2 || c2 == '?') {
                        res[i][j] = res[i+1][j+1];
                    } else if (c2 != '*') {
                        res[i][j] = false;
                    } else {
                        res[i][j] = res[i + 1][j] || res[i][j + 1];
                    }
                }
            }
        }

        return res[0][0];
    }
}
